src/main/java/com/bravo/user/dao/specification/AbstractSpecification.java

```diff
@@ -79,28 +80,57 @@ protected void applyDateTimeFilter(
     final List<String> inClause = new ArrayList<>();
 
     for(String s : values){
-      final boolean isLike = s.contains("%") || s.contains("*");
-      final boolean isNot = s.startsWith("!");
+      final boolean isLike = isLike(s);
+      final boolean isNot = isNot(s);
 
       if(!isLike && !isNot){
         inClause.add(s.toLowerCase());
         continue;
       }
       final String targetValue = (isNot ? s.substring(1) : s).toLowerCase();
-      Predicate predicate;
-      if(isLike){
-        predicate = criteriaBuilder.like(targetPath, targetValue.replace("*", "%"));
-      } else {
-        predicate = criteriaBuilder.equal(targetPath, targetValue);
-      }
-
-      if(isNot){
-        predicate = criteriaBuilder.not(predicate);
-      }
+      Predicate predicate = createPredicate(targetPath, targetValue, isLike, isNot);
       predicates.add(predicate);
     }
     if(!inClause.isEmpty()){
       applyInFilter(targetPath, inClause);
     }
   }
+
+  protected void applyStringFilterToFields(Collection<Expression<String>> fields, String value) {
+    final boolean isLike = isLike(value);
+    final boolean isNot = isNot(value);
+
+    final String targetValue = (isNot ? value.substring(1) : value).toLowerCase();
+    Predicate[] filterPredicates = fields.stream().map(field -> createPredicate(field, targetValue, isLike, isNot)).toArray(Predicate[]::new);
+
+    if(isNot){
+      predicates.add(criteriaBuilder.and(filterPredicates));
+    } else {
+      predicates.add(criteriaBuilder.or(filterPredicates));
+    }
```

Comment:
I know predicates is class-level.  Changing it on the fly like this makes this class hard to test.  Is there a better way (honestly, there may not be)?

Reply:
Every other method in the class adds predicates to this list which are then all anded together.  I'm not sure there is a much better way without refactoring the whole class.  In either event, tests could be accomplished by executing the methods and looking for the appropriate predicates in the list.